File: Medium/1315-Sum of Nodes with Even-Valued Grandparent/DFS.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sumEvenGrandparent(self, root: TreeNode) -> int:
        stack = [root]
        ret = 0

        while stack:
            curr = stack.pop()

            if curr.left:
                stack.append(curr.left)
                if curr.val % 2 == 0:
                    ret += curr.left.left.val if curr.left.left else 0
                    ret += curr.left.right.val if curr.left.right else 0
            if curr.right:
                stack.append(curr.right)
                if curr.val % 2 == 0:
                    ret += curr.right.left.val if curr.right.left else 0
                    ret += curr.right.right.val if curr.right.right else 0

        return ret

File: Problemset/0430-Flatten a Multilevel Doubly Linked List/DFS.rb
# Definition for a Node.
# class Node
#     attr_accessor :val, :prev, :next, :child
#     def initialize(val=nil, prev=nil, next_=nil, child=nil)
#         @val = val
#         @prev = prev
#         @next = next_
#         @child = child
#     end
# end

# @param {Node} root
# @return {Node}
def flatten(root)
    return nil if root.nil?

    prev = Node.new
    stack = [root]

    until stack.empty?
        curr = stack.pop

        stack.push(curr.next) unless curr.next.nil?
        stack.push(curr.child) unless curr.child.nil?

        prev.next = curr
        curr.prev = prev
        curr.child = nil

        prev = curr
    end

    root.prev = nil

    return root
end
